src/hotspot/cpu/aarch64/aarch64_neon_ad.m4

```diff
@@ -349,6 +349,29 @@ VECTOR_CAST_F2F(F, D, X, D, fcvtl, 2S, 2D)
 VECTOR_CAST_F2F(D, F, D, X, fcvtn, 2D, 2S)
 dnl
 
+define(`VECTOR_JAVA_FROUND', `
+instruct vround$7$2to$5$3($7 dst, $7 src, $7 tmp1, $7 tmp2, $7 tmp3)
+%{
+  predicate(n->as_Vector()->length() == $5 && n->bottom_type()->is_vect()->element_basic_type() == T_$6);
+  match(Set dst (RoundV$1 src));
+  effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2, TEMP tmp3);
+  format %{ "vround  $dst, $4, $src\t# round $7 $2 to $5$3 vector" %}
+  ins_encode %{
+    __ vector_round_neon(as_FloatRegister($dst$$reg), as_FloatRegister($src$$reg),
+                         as_FloatRegister($tmp1$$reg), as_FloatRegister($tmp2$$reg),
+                         as_FloatRegister($tmp3$$reg), __ $4);
+  %}
+  ins_pipe(pipe_class_default);
+%}')dnl           $1  $2  $3   $4 $5    $6    $7
+VECTOR_JAVA_FROUND(F, 2F,  I, T2S, 2,  INT, vecD)
+VECTOR_JAVA_FROUND(F, 4F,  I, T4S, 4,  INT, vecX)
+VECTOR_JAVA_FROUND(D, 2D,  L, T2D, 2, LONG, vecX)
+
+// Versions of vector_round for UseSVE > 0
+VECTOR_JAVA_FROUND(F, 2F,  I, T2S, 2,  INT, vReg)
+VECTOR_JAVA_FROUND(F, 4F,  I, T4S, 4,  INT, vReg)
+VECTOR_JAVA_FROUND(D, 2D,  L, T2D, 2, LONG, vReg)
+
```

Comment:
I don't know why do we need these rules. Should "UseSVE > 0" all go to the rules in sve ad file which call to vector_round_sve()?

Reply:
> I don't know why do we need these rules. Should "UseSVE > 0" all go to the rules in sve ad file which call to vector_round_sve()?

The freely-available Arm® Neoverse V1 Software Optimization Guide shows instructions such as ASIMD `FRINTA` having a throughput of 2 operations per clock, whereas it shows SVE `FRINTA` has a throughput of 1 operation per clock. This is true of most instructions used in `Math.round()`. I conclude that on V1, for short vectors, if we use ASIMD rather than equivalent SVE instructions, we should expect to virtually double throughput. For vectors wider than ASIMD supports, SVE should be a win.

At present, there is no reason not to use ASIMD for short vectors on all AArch64 processors. It won't significantly impair performance, and I can't think of any future circumstances in which it might.

Reply:
OK, thanks! Looks reasonable to me. We are going to make all vecX/vecD regs to vReg, I think that should make SIMD code cleaner.

Currently all rules for vReg are in aarch64_sve.ad. And since the codegen is actually for SVE target, though generates ASIMD insns, perhaps move these rules to aarch64_sve.ad would be better? Also I think the 2F/4F rules could be merged into one, like:

```
instruct vroundvRegF(vReg dst, vReg src, vReg tmp1, vReg tmp2, vReg tmp3)
%{
  predicate(n->as_Vector()->length_in_bytes() <= 16);
  match(Set dst (RoundVF src));
  effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2, TEMP tmp3);
  format %{ "vround  $dst, $src\t# round vReg F to I vector" %}
  ins_encode %{
    uint size = Matcher::vector_length_in_bytes(this);
    __ vector_round_neon(as_FloatRegister($dst$$reg), as_FloatRegister($src$$reg),
                         as_FloatRegister($tmp1$$reg), as_FloatRegister($tmp2$$reg),
                         as_FloatRegister($tmp3$$reg), (size == 16) ? __ T4S : __ T2S);
  %}
  ins_pipe(pipe_slow);
%}
```

Reply:
Seems reasonable. Maybe we could the logic down into MacroAssembler. That way there'd be one point at which the SVE/Neon devcision was made, in MacroAssembler. The disadvantage would be that Neon and SVE versions require different register clobbers, but that might not matter.

Reply:
Hi, do you like it better now? Thanks.

Reply:
> Hi, do you like it better now? Thanks.

Yeah, looks good to me now! And the jmh results on some of our test systems (with and without SVE) look quite good.

Reply:
That's great. Thanks for helping by testing the patch, I appreciate it.